src/net/sourceforge/kolmafia/textui/Line.java

```diff
@@ -162,13 +165,19 @@ private Token(final int tokenLength) {
         // Going forward, we can just assume lineRemainder is an
         // empty string.
         lineRemainder = "";
+        tokenLength = 0;
       } else {
         final String lineRemainderWithToken = Line.this.substring(offset);
 
         this.content = lineRemainderWithToken.substring(0, tokenLength);
         lineRemainder = lineRemainderWithToken.substring(tokenLength);
       }
 
+      // 0-indexed line
+      final int lineNumber = Math.max(0, Line.this.lineNumber - 1);
```

Comment:
Tokens' line are stored as 0-indexed (as required by the Position class and LSP4J), but Lines store their number as 1-index because of how we present them to the user as 1-indexed (which is a good thing).

If we were to store them as 0-indexed, all we would do is move "when" the increment happens.

Is it possible? Yes.
Will it make it so that the code requires one less operation? No, it'll add an additional one.

Reply:
The operation I'd really like to get rid of is Math.max(), since it indicates that an abstraction isn't consistent (namely, "Line.lineNumber is 1-indexed").

I believe the only place where Line.lineNumber is set to 0 is if Line.content is null (i.e. at end-of-file) and Line.previousLine is null (namely, when the script is an empty file).

Reply:
Reasonable.

Reply:
@heeheehee-kolmafia this doesn't work. It doesn't get rid of the Math.max().

The line number still gets incremented one last time before reaching the "end of file", which we need to decrement.

Reply:
What IS the culprit, however, is *this* line
```java
      this.lineNumber = this.previousLine != null ? this.previousLine.lineNumber : 0;
```
->
```java
      this.lineNumber = this.previousLine != null ? this.previousLine.lineNumber : 1;
```

Reply:
(It was meant as an indicator that "there was absolutely nothing to read", but I guess it's not really needed; we can just see that it doesn't even have a previousLine)

Reply:
Yeah, hence my comment that this was the only place where lineNumber was set to 0 (Line.content == null && Line.previousLine == null).